src/client/datascience/interactive-window/nativeInteractiveWindowProvider.ts

```diff
@@ -109,69 +110,83 @@ export class NativeInteractiveWindowProvider implements IInteractiveWindowProvid
+                const result = new NativeInteractiveWindow(
+                    this.serviceContainer.get<IApplicationShell>(IApplicationShell),
+                    this.serviceContainer.get<IDocumentManager>(IDocumentManager),
+                    this.serviceContainer.get<IStatusProvider>(IStatusProvider),
+                    this.serviceContainer.get<IFileSystem>(IFileSystem),
+                    this.serviceContainer.get<IConfigurationService>(IConfigurationService),
+                    this.serviceContainer.get<ICommandManager>(ICommandManager),
+                    this.serviceContainer.get<INotebookExporter>(INotebookExporter),
+                    this.serviceContainer.get<IWorkspaceService>(IWorkspaceService),
+                    resource,
+                    mode,
+                    this.serviceContainer.get<IPythonExtensionChecker>(IPythonExtensionChecker),
+                    this.serviceContainer.get<IExportDialog>(IExportDialog),
+                    notebookDocument,
+                    this.notebookControllerManager,
+                    this.kernelProvider,
+                    this.disposables,
+                    this.serviceContainer.get<IJupyterDebugger>(IJupyterDebugger)
+                );
+                this._windows.push(result);
 
-        // This is the last interactive window at the moment (as we're about to create it)
-        this.lastActiveInteractiveWindow = result;
+                // This is the last interactive window at the moment (as we're about to create it)
+                this.lastActiveInteractiveWindow = result;
 
-        // When shutting down, we fire an event
-        const handler = result.closed(this.onInteractiveWindowClosed);
-        this.disposables.push(result);
-        this.disposables.push(handler);
-        this.disposables.push(result.onDidChangeViewState(this.raiseOnDidChangeActiveInteractiveWindow.bind(this)));
+                // When shutting down, we fire an event
+                const handler = result.closed(this.onInteractiveWindowClosed);
+                this.disposables.push(result);
+                this.disposables.push(handler);
+                this.disposables.push(
+                    result.onDidChangeViewState(this.raiseOnDidChangeActiveInteractiveWindow.bind(this))
+                );
 
-        // Show in the background
-        result.show().ignoreErrors();
+                // fire created event
+                this._onDidCreateInteractiveWindow.fire(result);
+                return result;
+            });
+    }
 
-        // fire created event
-        this._onDidCreateInteractiveWindow.fire(result);
+    protected async create(resource: Resource, mode: InteractiveWindowMode): Promise<NativeInteractiveWindow> {
+        // If there's no resource, just create a new one
+        if (resource === undefined) {
+            return this.createInteractiveWindowPromise(resource, mode);
```

Comment:
Could this cause the same race? You might create two interactive windows?

Wondering if the undefined case should also have an entry in the map

Reply:
If resource is undefined that means there's no backing script file, i.e. being created from command palette, in which case we would always want to create a new interactive window, no?

Reply:
Looking at the old code, it was async too, but the creation of the window was based entirely on the mode. Why does the controller need to be created before the window?

Reply:
The old code would just await the async parts after creation. Maybe we can live chat instead?